pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/bucket/BookkeeperBucketSnapshotStorage.java

```diff
@@ -0,0 +1,243 @@
+        return createLedger(bucketKey)
+                .thenCompose(ledgerHandle -> addEntry(ledgerHandle, snapshotMetadata.toByteArray())
+                        .thenCompose(__ -> addSnapshotSegments(ledgerHandle, bucketSnapshotSegments))
+                        .thenCompose(__ -> closeLedger(ledgerHandle))
+                        .thenApply(__ -> ledgerHandle.getId()));
+    }
+
+    @Override
+    public CompletableFuture<SnapshotMetadata> getBucketSnapshotMetadata(long bucketId) {
+        return openLedger(bucketId).thenCompose(
+                ledgerHandle -> getLedgerEntryThenCloseLedger(ledgerHandle, 0, 0).
+                        thenApply(entryEnumeration -> parseSnapshotMetadataEntry(entryEnumeration.nextElement())));
+    }
+
+    @Override
+    public CompletableFuture<List<SnapshotSegment>> getBucketSnapshotSegment(long bucketId, long firstSegmentEntryId,
+                                                                             long lastSegmentEntryId) {
+        return openLedger(bucketId).thenCompose(
+                ledgerHandle -> getLedgerEntryThenCloseLedger(ledgerHandle, firstSegmentEntryId,
+                        lastSegmentEntryId).thenApply(this::parseSnapshotSegmentEntries));
+    }
+
+    @Override
+    public CompletableFuture<Long> getBucketSnapshotLength(long bucketId) {
+        return openLedger(bucketId).thenApply(ledgerHandle -> {
+            long length = ledgerHandle.getLength();
+            closeLedger(ledgerHandle);
+            return length;
+        });
+    }
+
+    @Override
+    public CompletableFuture<Void> deleteBucketSnapshot(long bucketId) {
+        return deleteLedger(bucketId);
+    }
+
+    @Override
+    public void start() throws Exception {
+        pulsar.getBookKeeperClient();
+        this.bookKeeper = pulsar.getBookKeeperClientFactory().create(
+                pulsar.getConfiguration(),
+                pulsar.getLocalMetadataStore(),
+                pulsar.getIoEventLoopGroup(),
+                Optional.empty(),
+                null
+        );
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (bookKeeper != null) {
+            bookKeeper.close();
+        }
+    }
+
+    private CompletableFuture<Void> addSnapshotSegments(LedgerHandle ledgerHandle,
+                                                        List<SnapshotSegment> bucketSnapshotSegments) {
+        List<CompletableFuture<Void>> addFutures = new ArrayList<>();
+        for (SnapshotSegment bucketSnapshotSegment : bucketSnapshotSegments) {
+            addFutures.add(addEntry(ledgerHandle, bucketSnapshotSegment.toByteArray()));
+        }
+
+        return FutureUtil.waitForAll(addFutures);
+    }
+
+    private SnapshotMetadata parseSnapshotMetadataEntry(LedgerEntry ledgerEntry) {
+        try {
+            return SnapshotMetadata.parseFrom(ledgerEntry.getEntry());
+        } catch (InvalidProtocolBufferException e) {
+            throw new RuntimeException(e);
```

Comment:
Why not declare a specified exception for the snapshot storage? The explicit exception would help us to decide how to handle them better in the next step.